### ISSUE TITLE
DNS resolution failing on macsys after upgrading to 1.39.x

### ISSUE THREAD

**Issue by lincolnmantracer** [tailscale/tailscale#7769]
### What is the issue?

I recently upgraded Tailscale macsys from 1.38.3 to 1.39.x (.2 I think?). DNS resolution immediately started failing despite several restarts and "reinstalling" the system extension. Errors from a go program looked like this:

```
lookup foo.example.com on 100.100.100.100:53: read udp 100.124.137.53:63265->100.100.100.100:53: i/o timeout
```

We use split dns for the domain in question, with "override local DNS" checked. Downgrading to 1.38 immediately fixed the issue.

### OS

macOS

### OS version

macOS Monterey 12.3.1

### Tailscale version

1.39.2

### Bug report

BUG-beeb61f2c7675c37eed3fc1620c04be9cbcfc4a05ea7d8675daad3110d57edd0-20230403123654Z-b7b311863d49a81e

**Comment by mihaip** [comment]
Looking at our [internal logs](http://logz.ts.tailscale.com:9431/logs?start=2023-03-10T16%3A58%3A20Z&end=2023-04-10T16%3A58%3A20Z&instances=beeb61f2c7675c37eed3fc1620c04be9cbcfc4a05ea7d8675daad3110d57edd0&match=goSetupLogs), these are the versions that were installed on that machine:
- 1.39.55 (installed on March 27)
- 1.39.92 (installed on April 2)
- 1.38.3 (installed on April 3)

Based on how broken things are, I'm guessing that this started to happen with the 1.39.92 release rather than the 1.39.55 one, which narrows down the regression range somewhat.

Changes between 1.39.55 and 1.39.92: [open source](https://github.com/tailscale/tailscale/compare/4a89642f7...7b4e85aa7), [closed source](https://github.com/tailscale/corp/compare/52f5a6d7f...b07177ba9). Nothing immediately jumps out.

**Comment by mihaip** [comment]
@knyar is also experiencing this, and has narrowed it down a bit: 1.39.73 is the first version it’s broken in; 1.39.55 works fine ([open source changes](https://github.com/tailscale/tailscale/compare/4a89642f7...920ec6924), [closed source changes](https://github.com/tailscale/corp/compare/52f5a6d7f...d55f94b15)).

He also points out that a new error message has appeared during startup ([logs](http://logz.ts.tailscale.com:9431/logs?start=2023-03-10T16%3A58%3A20Z&end=2023-04-10T16%3A58%3A20Z&instances=beeb61f2c7675c37eed3fc1620c04be9cbcfc4a05ea7d8675daad3110d57edd0&match=%28goSetupLogs%7CpeerAPI+listen%29)):

```
14:23:21 | goSetupLogs: success (started with NaNM); version 1.39.92-t7b4e85aa7-gb07177ba9, go1.20.1-ts568add9
14:23:26 | peerapi: failed to do peerAPI listen, harmless (netstack available) but error was: listen tcp4 100.124.137.53:0: bind: can't assign requested address
14:23:26 | peerapi: failed to do peerAPI listen, harmless (netstack available) but error was: listen tcp6 [fd7a:115c:a1e0:ab12:4843:cd96:627c:8935]:0: bind: can't assign requested address
```

While the peerAPI listener should not be involved in DNS requests, it does show that something changed.

**Comment by mihaip** [comment]
One possible change is that @danderson changed how we build the Mac app in tailscale/corp#10078. The build tags used to be specified in [a Redo file](https://github.com/tailscale/corp/blob/72401f9d3bd6eecfde5f5a2cb70a5469fb012bf0/xcode/ipn-go-bridge/default.a.do#L45):

```
export BUILD_TAGS="-tags=tailscale_go,ts_macext,omitidna,omitpemdecrypt"
```

And are now done via `gocross`: https://github.com/tailscale/tailscale/blob/33b006cacf84081ebda268d5241cb9e7bdc4ff9d/tool/gocross/autoflags.go#L101-L135

Specifically the `ts_macext` tag is no longer being passed in, and it controls the behavior of various bits of code, including DNS resolution: https://github.com/tailscale/tailscale/blob/33b006cacf84081ebda268d5241cb9e7bdc4ff9d/net/dns/resolver/macios_ext.go#L4

Will look into adding it back.

**Comment by mihaip** [comment]
This should be fixed with 1.39.146.
